backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/invoicecandidate/spi/impl/ProductFromOrderToInvoiceCandidateTest.java

```diff
@@ -0,0 +1,268 @@
+		bpLocation.setC_BPartner_ID(bpartner.getC_BPartner_ID());
+		save(bpLocation);
+
+		bpartnerAndLocationId = BPartnerLocationId.ofRepoId(bpLocation.getC_BPartner_ID(), bpLocation.getC_BPartner_Location_ID());
+	}
+
+	private void initHandlers()
+	{
+		handler = create(ctx, I_C_ILCandHandler.class, trxName);
+		handler.setC_ILCandHandler_ID(540001);
+		handler.setClassname(C_OrderLine_Handler.class.getName());
+		handler.setName("Auftragszeilen");
+		handler.setTableName(I_C_OrderLine.Table_Name);
+		save(handler);
+		olHandler.setHandlerRecord(handler);
+	}
+
+	private void initC_DocType()
+	{
+		docType = create(ctx, I_C_DocType.class, trxName);
+		docType.setAD_Org_ID(orgId.getRepoId());
+		docType.setC_DocType_ID(1000016);
+		save(docType);
+	}
+
+	private void initC_Order()
+	{
+		order = create(ctx, I_C_Order.class, trxName);
+		order.setAD_Org_ID(orgId.getRepoId());
+		order.setC_BPartner_ID(bpartnerAndLocationId.getBpartnerId().getRepoId());
+		order.setC_BPartner_Location_ID(bpartnerAndLocationId.getRepoId());
+		order.setBill_BPartner_ID(bpartnerAndLocationId.getBpartnerId().getRepoId());
+		order.setBill_Location_ID(bpartnerAndLocationId.getRepoId());
+		order.setDocStatus(DocStatus.Completed.getCode());
+		order.setInvoiceRule(InvoiceRule.AfterDelivery.getCode());
+		order.setC_DocTypeTarget_ID(1000016);
+		order.setDatePromised(Timestamp.valueOf("2021-11-30 00:00:00"));
+		order.setC_Currency_ID(10);
+		order.setM_PricingSystem_ID(20);
+		save(order);
+	}
+
+	private void initC_OrderLine()
+	{
+		orderLine = create(ctx, I_C_OrderLine.class, trxName);
+
+		orderLine.setAD_Org_ID(orgId.getRepoId());
+		orderLine.setM_Product_ID(productId.getRepoId());
+
+		final BigDecimal qty = new BigDecimal(13);
+		orderLine.setQtyEntered(qty);
+		orderLine.setQtyOrdered(qty);
+		orderLine.setC_UOM_ID(stockUomId.getRepoId());
+		orderLine.setQtyReserved(qty);
+		orderLine.setQtyDelivered(qty);
+		orderLine.setQtyInvoiced(qty);
+		orderLine.setC_Order_ID(order.getC_Order_ID());
+		orderLine.setC_Tax_ID(createTax().getRepoId());
+
+		save(orderLine);
+	}
+
+	private TaxId createTax()
+	{
+		final I_C_Tax tax = newInstance(I_C_Tax.class);
+		saveRecord(tax);
+
+		return TaxId.ofRepoId(tax.getC_Tax_ID());
+	}
+
+
+
+	@Test
+	public void testProductChanged()
+	{
+		final List<I_C_Invoice_Candidate> invoiceCandidates = olHandler
+				.createCandidatesFor(InvoiceCandidateGenerateRequest.of(olHandler, orderLine))
+				.getC_Invoice_Candidates();
+		assertThat(invoiceCandidates).as("There is 1 invoice candidate").hasSize(1);
+
+		final I_C_Invoice_Candidate candidate = invoiceCandidates.get(0);
+		assertThat(candidate.getInvoiceRule()).isNotNull();
+		saveRecord(candidate);
+		// ReActivate
+		docActionBL.processEx(order, IDocument.ACTION_ReActivate, IDocument.STATUS_InProgress);
+		orderLine.setM_Product_ID(changedProductId.getRepoId());
+
+		olHandler.setOrderedData(candidate);
```

Comment:
TBH if u call setOrderedData directly here.... i am not sure what are you exactly testing....

i.e. u could call that method directly without reactivating the order....

so what's exactly ur test scenario?

Reply:
- Like the WEB UI test:

1. Create an SO and complete it
2. IC is created with product 1
3. Reactivate the SO
4. Change the product in the OL
5. Compare the product in the OL and the IC